src/graph/executor/algo/AllPathsExecutor.cpp

```diff
@@ -462,6 +353,241 @@ AllPathsExecutor::doBuildPath(size_t step,
       });
 }
 
+std::vector<Value> AllPathsExecutor::convertNPath2List(NPath* path, bool reverse) {
+  std::vector<Value> list;
+  NPath* head = path;
+  while (head != nullptr) {
+    list.emplace_back(head->edge);
+    list.emplace_back(head->vertex);
+    head = head->p;
+  }
+  if (!reverse) {
+    std::reverse(list.begin(), list.end());
+  }
+  return list;
+}
+
+void AllPathsExecutor::buildOneWayPath(std::vector<NPath*>& paths, bool reverse) {
+  auto& initVids = reverse ? rightInitVids_ : leftInitVids_;
+  for (auto& path : paths) {
+    auto& edgeVal = path->edge;
+    auto& dst = edgeVal.getEdge().dst;
+    auto findDst = initVids.find(dst);
+    if (findDst == initVids.end()) {
+      continue;
+    }
+    ++cnt_;
+    if (cnt_.load(std::memory_order_relaxed) > limit_) {
+      break;
+    }
+    auto valueList = convertNPath2List(path, !reverse);
+    Row row;
+    Value emptyPropVertex(Vertex(dst, {}));
+    if (!reverse) {
+      auto dstVertex = valueList.back();
+      valueList.pop_back();
+      // add src
+      row.values.emplace_back(emptyPropVertex);
+      List edgeList(std::move(valueList));
+      // add edgeList
+      row.values.emplace_back(std::move(edgeList));
+      // add dst
+      row.values.emplace_back(dstVertex);
+    } else {
+      // add src
+      row.values.emplace_back(valueList.front());
+      std::vector<Value> tmp(std::make_move_iterator(valueList.begin() + 1),
+                             std::make_move_iterator(valueList.end()));
+      List edgeList(std::move(tmp));
+      // add edgeList
+      row.values.emplace_back(std::move(edgeList));
+      // add dst
+      row.values.emplace_back(emptyPropVertex);
+    }
+    auto iter = emptyPropVertices_.find(emptyPropVertex);
+    if (iter == emptyPropVertices_.end()) {
+      emptyPropVids_.emplace_back(dst);
+      emptyPropVertices_.emplace(emptyPropVertex);
+    }
+    result_.rows.emplace_back(std::move(row));
+  }
+}
+
+std::vector<Row> AllPathsExecutor::buildOneWayPathFromHashTable(bool reverse) {
+  auto& initVids = reverse ? rightInitVids_ : leftInitVids_;
+  std::vector<Row> result;
+  for (auto& vid : initVids) {
+    auto findVid = hashTable_.find(vid);
+    if (findVid == hashTable_.end()) {
+      continue;
+    }
+    auto& paths = findVid->second;
+    Value emptyPropVertex(Vertex(vid, {}));
+    if (!reverse) {
+      for (auto& path : paths) {
+        ++cnt_;
```

Comment:
defalt ++ is memory_order_seq_cst, is it what you want, or here we can use memory_order_relaxed?